Allow tuning red color sensitivity of infra-red distance sensor

jmarsik opened the pull request:
**Description**
Allows tuning of red color sensitivity of infra-red distance sensor with a new property `redColorSensitivity`. It's a floating point number, with default value 1 it will behave like before, with value 0 the red color sensitivity will be completely disabled.

**Tasks**
  - [x] Add to changelog
  - [x] Add entry to field summary documentation

**Documentation**
https://cyberbotics.com/doc/reference/distancesensor?version=develop

**Additional context**
This is my first PR to Webots and I may not know all the things that have to be done.
Should I update the changelog file?
Should I include controller method to get the new property (`wb_distance_sensor_get_red_color_sensitivity`)?

jmarsik commented:
Or should I base this on the master branch to speed up inclusion in Webots release version? I have chosen develop as the base but I see other pull requests that are in fact new features / enhancements and are merged into master.

omichel commented on src/webots/nodes/WbDistanceSensor.cpp:
```diff
@@ -619,6 +623,7 @@ void WbDistanceSensor::addConfigure(QDataStream &stream) {
   stream << (double)mLut->minValue();
   stream << (double)mLut->maxValue();
   stream << (double)mAperture->value();
+  stream << (double)mRedColorSensitivity->value();
```
That seems wrong as the red color sensitivity should not be sent to the controller (and it's not parsed by the libController).
```suggestion
```

omichel reviewed (changes requested):
I believe it's not needed to add an API function to retrieve the red color sensitivity. However, a line in the change log is needed.

omichel commented:
> Or should I base this on the master branch to speed up inclusion in Webots release version? I have chosen develop as the base but I see other pull requests that are in fact new features / enhancements and are merged into master.

Since this is a new feature, it is correct to target the `develop` branch. Anyhow, we don't plan to release any R2021a revision, and R2021b will be based on the `develop` branch, so your contribution will reach the release very soon (normally end of June).

ad-daniel commented:
Since you're adding a new field to the `DistanceSensor` you should also mention and describe it in the [field summary](https://cyberbotics.com/doc/reference/distancesensor#field-summary) of the documentation as it's done for the others. Probably what you added to the `type` field documentation can be moved there afterwards as well.

Also for the documentation, generally we write one phrase per line (so you should go to a new line every period) to avoid having to fix formatting issues.

Thank you for the contribution!

jmarsik commented:
I have added changelog entry, added field summary entry into the documentation of the distance sensor and removed wrong addition of the field to controller stream data. I have also fixed the calculation itself because it was wrong, sorry for that.

Regarding the June release - that would be great!

Thank you all.

omichel commented on docs/reference/changelog-r2021.md:
```diff
@@ -8,7 +8,9 @@ Released on June, Xth, 2021.
     - Added the `wb_supervisor_node_save/load_state` functions that allow partial world reverting to a saved state ([#2740](https://github.com/cyberbotics/webots/pull/2740)).
     - Added coupled motors feature which allows to control multiple logically linked [Motors](motor.md) at once ([#2939](https://github.com/cyberbotics/webots/pull/2939)).
-    - Changed the rendering engine of the streaming viewer and of the animations for WREN ([#2769](https://github.com/cyberbotics/webots/pull/2769)).
+    - Changed the rendering engine of the streaming viewer and of the animations for WREN ([#2769](https://
+    github.com/cyberbotics/webots/pull/2769)).
```
```suggestion
    - Changed the rendering engine of the streaming viewer and of the animations for WREN ([#2769](https://github.com/cyberbotics/webots/pull/2769)).
```

omichel commented on docs/reference/changelog-r2021.md:
```diff
@@ -8,7 +8,9 @@ Released on June, Xth, 2021.
-    - Changed the rendering engine of the streaming viewer and of the animations for WREN ([#2769](https://github.com/cyberbotics/webots/pull/2769)).
+    - Changed the rendering engine of the streaming viewer and of the animations for WREN ([#2769](https://
+    github.com/cyberbotics/webots/pull/2769)).
+    - Added the `redColorSensitivity` field to [DistanceSensor](distancesensor.md) that allows tuning (or even disabling) of the red color sensitivity for an infra-red distance sensor type ([#3077](https://github.com/cyberbotics/webots/pull/3077)).
```
```suggestion
    - Added the `redColorSensitivity` field to [DistanceSensor](distancesensor.md) that allows tuning (or even disabling) of the red color sensitivity for an infra-red distance sensor ([#3077](https://github.com/cyberbotics/webots/pull/3077)).
```

omichel commented on docs/reference/distancesensor.md:
```diff
@@ -76,6 +77,13 @@ This field is ignored for the "sonar" and "laser" DistanceSensor types.
 
+- `redColorSensitivity`: red color sensitivity factor.
+This allows to tune (or even disable) red color sensitivity for infra-red distance sensor type.
+Value of 1 means ordinary Webots behavior.
```
```suggestion
A value of 1 corresponds to the default behavior.
```

omichel commented on docs/reference/distancesensor.md:
```diff
@@ -76,6 +77,13 @@ This field is ignored for the "sonar" and "laser" DistanceSensor types.
+This allows to tune (or even disable) red color sensitivity for infra-red distance sensor type.
+Value of 1 means ordinary Webots behavior.
+Values greater that 1 increase the red color sensitivity and values lower than 1 decrease it.
+Value of 0 disables the red color sensitivity completely.
```
```suggestion
A value of 0 disables the effect of the red color completely.
```

omichel commented on docs/reference/distancesensor.md:
```diff
@@ -4,12 +4,13 @@ Derived from [Device](device.md) and [Solid](solid.md).
+  SFFloat  aperture                 1.5708                # [0, 2*pi]
+  SFFloat  gaussianWidth            1                     # [0, inf)
+  SFFloat  resolution               -1                    # {-1, [0, inf)}
+  SFFloat  redColorSensitivity      1                     # [0, 1]
```
That seems more correct:
```suggestion
  SFFloat  redColorSensitivity      1                     # [0, inf)
```

jmarsik replied:
Actually even value bigger than 1 will work, it will increase the red color sensitivity of the sensor above the original Webots behavior. But we can go with 0-1, no problem. I even mention it in the WRL file. Maybe I should add a validation then?

omichel reviewed (changes requested):
We are almost there...

omichel commented on resources/nodes/DistanceSensor.wrl:
```diff
@@ -21,12 +21,13 @@ DistanceSensor {
+  field     SFFloat                                             aperture             1.5708                 # aperture angle
+  field     SFFloat                                             gaussianWidth        1                      # gaussian distribution width
+  field     SFFloat                                             resolution          -1                      # resolution
+  field     SFFloat                                             redColorSensitivity  1                      # red color sensitivity (0..1) for infra-red distance sensor
```
I forgot this "(0..1)" which doesn't seems consistent with the documentation now and is misleading here.
```suggestion
  field     SFFloat                                             redColorSensitivity  1                      # red color sensitivity for infra-red distance sensor
```

omichel reviewed (changes requested):
Oops, a last one 😉.

omichel reviewed (approved):
Thank you.

jmarsik commented:
> Thank you.

No problem. Thank you for really quick processing of this pull request!

jmarsik commented:
Now I will squash merge this PR, is that ok? The button is now available, after your approval and adding me to the commiters team.

omichel commented:
> Now I will squash merge this PR, is that ok? The button is now available, after your approval and adding me to the commiters team.

Yes, you can go ahead and squash and merge.
